src/D3MHub.sol

```diff
@@ -0,0 +1,603 @@
+
+        if (mode == Mode.NORMAL) {
+            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
+            vat.slip(ilk, address(_pool), -int256(amount));
+            vat.move(address(this), vow, fees * RAY);
+        } else if (mode == Mode.MODULE_CULLED) {
+            vat.slip(ilk, address(_pool), -int256(amount));
+            vat.move(address(this), vow, total * RAY);
+        } else {
+            // This can be done with the assumption that the price of 1 collateral unit equals 1 DAI.
+            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
+            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
+            vat.slip(ilk, address(_end), -int256(amount));
+            vat.move(address(this), vow, total * RAY);
+        }
+
+        emit Unwind(ilk, amount, fees);
+    }
+
+    // Ilk Getters
+    /**
+        @notice Return pool of an ilk
+        @param ilk   bytes32 of the D3M ilk
+        @return pool address of pool contract
+    */
+    function pool(bytes32 ilk) external view returns (address) {
+        return address(ilks[ilk].pool);
+    }
+
+    /**
+        @notice Return plan of an ilk
+        @param ilk   bytes32 of the D3M ilk
+        @return plan address of plan contract
+    */
+    function plan(bytes32 ilk) external view returns (address) {
+        return address(ilks[ilk].plan);
+    }
+
+    /**
+        @notice Return tau of an ilk
+        @param ilk  bytes32 of the D3M ilk
+        @return tau sec until debt can be written off
+    */
+    function tau(bytes32 ilk) external view returns (uint256) {
+        return ilks[ilk].tau;
+    }
+
+    /**
+        @notice Return culled status of an ilk
+        @param ilk  bytes32 of the D3M ilk
+        @return culled whether or not the d3m has been culled
+    */
+    function culled(bytes32 ilk) external view returns (uint256) {
+        return ilks[ilk].culled;
+    }
+
+    /**
+        @notice Return tic of an ilk
+        @param ilk  bytes32 of the D3M ilk
+        @return tic timestamp of when d3m is caged
+    */
+    function tic(bytes32 ilk) external view returns (uint256) {
+        return ilks[ilk].tic;
+    }
+
+    /**
+        @notice Main function for updating a D3M position.
+        Determines the current state and either winds or unwinds as necessary.
+        @dev Winding the target position will be constrained by the Ilk debt
+        ceiling, the overall DSS debt ceiling and the maximum deposit by the
+        pool. Unwinding the target position will be constrained by the number
+        of assets available to be withdrawn from the pool.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function exec(bytes32 ilk) external lock {
+        (uint256 Art, uint256 rate,, uint256 line,) = vat.ilks(ilk);
+        require(rate == RAY, "D3MHub/rate-not-one");
```

Comment:
I think we could also check that `spot` equals to `RAY`. We are already bringing this variable and a different value would break the normal behavior of this module in a similar way.

Reply:
7354513